TUM Campus App/CardViewController.swift

```diff
@@ -109,38 +87,101 @@ extension CardViewController {
             self.tableView.tableHeaderView = search?.searchBar
         }
     }
+    
+}
+
+extension CardViewController: EditCardsViewControllerDelegate {
+    
+    func refresh(_ sender: AnyObject?) {
+        manager?.loadCards(skipCache: sender != nil).onSuccess(in: .main) { categories in
+            // TODO:
+            self.cards = categories.flatMap{ $0.elements.first }
+            self.categories = categories
+            self.tableView.reloadData()
+            self.refresh.endRefreshing()
+        }
+    }
+
+    func didUpdateCards() {
+        refresh(nil)
+        tableView.reloadData()
+    }
+}
+
+
+extension CardViewController: DetailViewDelegate {
+    
+    func dataManager() -> TumDataManager? {
+        return manager
+    }
+    
 }
 
-extension CardViewController {
+
+extension CardViewController: UITableViewDelegate, UITableViewDataSource {
     
-    override func numberOfSections(in tableView: UITableView) -> Int {
+    func numberOfSections(in tableView: UITableView) -> Int {
         return 1
     }
     
-    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
         return max(cards.count, 1)
     }
     
-    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
         return UITableViewAutomaticDimension
     }
     
-    override func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
+    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
         return 480
     }
     
-    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
         let item = cards | indexPath.row ?? EmptyCard()
+        
         let cell = tableView.dequeueReusableCell(withIdentifier: item.getCellIdentifier()) as? CardTableViewCell ?? CardTableViewCell()
-        cell.setElement(item)
+        
+        if let cell = cell as? SingleDataElementPresentable {
+             cell.setElement(item)
+        }
+        if let cell = cell as? MultipleDataElementsPresentable {
```

Comment:
Do you see a case where both would be happening at the same time? If not make it an `else if`

Reply:
changed